daemon/mgr/cri.go

```diff
@@ -695,55 +695,73 @@ func (c *CriManager) UpdateContainerResources(ctx context.Context, r *runtime.Up
 // ExecSync executes a command in the container, and returns the stdout output.
 // If command exits with a non-zero exit code, an error is returned.
 func (c *CriManager) ExecSync(ctx context.Context, r *runtime.ExecSyncRequest) (*runtime.ExecSyncResponse, error) {
-	// TODO: handle timeout.
 	id := r.GetContainerId()
 
+	timeout := time.Duration(r.GetTimeout()) * time.Second
+	var timeoutCh <-chan time.Time
+	if timeout == 0 {
+		// Do not set timeout if it's 0.
+		timeoutCh = make(chan time.Time)
+	} else {
+		timeoutCh = time.After(timeout)
+	}
+
 	createConfig := &apitypes.ExecCreateConfig{
 		Cmd: r.GetCmd(),
 	}
-
 	execid, err := c.ContainerMgr.CreateExec(ctx, id, createConfig)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create exec for container %q: %v", id, err)
 	}
 
-	var output bytes.Buffer
-	startConfig := &apitypes.ExecStartConfig{}
+	reader, writer := io.Pipe()
+	defer writer.Close()
+
 	attachConfig := &AttachConfig{
 		Stdout:      true,
 		Stderr:      true,
-		MemBuffer:   &output,
+		MemBuffer:   writer,
 		MuxDisabled: true,
 	}
 
+	startConfig := &apitypes.ExecStartConfig{}
+
 	err = c.ContainerMgr.StartExec(ctx, execid, startConfig, attachConfig)
 	if err != nil {
 		return nil, fmt.Errorf("failed to start exec for container %q: %v", id, err)
 	}
 
-	var execConfig *ContainerExecConfig
-	for {
-		execConfig, err = c.ContainerMgr.GetExecConfig(ctx, execid)
+	readWaitCh := make(chan error, 1)
+	var recv bytes.Buffer
+	go func() {
+		defer reader.Close()
+		_, err = io.Copy(&recv, reader)
+		if err != nil {
+			logrus.Infof("failed to read data from the pipe")
+		}
+		readWaitCh <- err
+	}()
+
+	select {
```

Comment:
Please add the `ctx.Done()` selector here because the request maybe canceled without waiting the result.

Reply:
```timeoutCh``` is doing the same thing as ctx.Done( ), since I have tried ctx.Done() and some unknown errors will be brought about by that.
And this is the exact way that cri-containerd used to check the ```timeout``` moment, therefore I am following the way it is.

Reply:
Actually, the `ctx.Done()` has been handled by the interceptor. It has the functionality like the `timeoutCh`, but not the same thing. Therefore, could you mind to provide information about what **unknown** error you mentioned?